update GetDataHandlers

ATFGK opened the pull request:
Fixes #2804

ATFGK committed:
Update GetDataHandlers.cs

ATFGK committed:
Update HandleProjectileNew

hakusaro commented:
This doesn't compile.

hakusaro reviewed (changes requested):
Please fix compilation errors / etc before a real review occurs

ATFGK committed:
update

punchready commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2912,17 +2914,33 @@ private static bool HandleProjectileNew(GetDataHandlerArgs args)
+			int origDmg = (int)(bitsByte[6] ? args.Data.ReadInt16() : 0);
+			int projUUID = (int)(bitsByte[7] ? args.Data.ReadInt16() : -1);
+			if (projUUID >= 1000) projUUID = -1;
+			ai[3] = (bitsByte2[0] ? args.Data.ReadSingle() : 0f);
```
index should be 2 here, not 3

punchready commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2912,17 +2914,33 @@ private static bool HandleProjectileNew(GetDataHandlerArgs args)
-			NewProjectileData bits = new NewProjectileData((BitsByte)args.Data.ReadByte());
+			
+			BitsByte bitsByte = args.Data.ReadByte();
+		        BitsByte bitsByte2 = bitsByte[2] ? args.Data.ReadByte() : 0;
```
`ReadByte` returns an `int`, not `byte`, you need to cast it first to be able to implicitly convert to `BitsByte`

punchready commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2912,17 +2914,33 @@ private static bool HandleProjectileNew(GetDataHandlerArgs args)
+			for (int i = 0; i < Projectile.maxAI; ++i) ai[i] = 0f;
+			ai[0] = (bitsByte[0] ? args.Data.ReadSingle() : 0f);
+			ai[1] = (bitsByte[1] ? args.Data.ReadSingle() : 0f);
+			int bannerId = (int)(bitsByte[3] ? args.Data.ReadUInt16() : 0);
+			int dmg = (int)(bitsByte[4] ? args.Data.ReadInt16() : 0);
+			float knockback = bitsByte[5] ? args.Data.ReadSingle() : 0f;
+			int origDmg = (int)(bitsByte[6] ? args.Data.ReadInt16() : 0);
+			int projUUID = (int)(bitsByte[7] ? args.Data.ReadInt16() : -1);
```
i would keep the old data types here instead of casting them all to int, no need to deviate from the previous code, especially because `dmg` actually needs to stay a `short` due to the `OnNewProjectile` hook storing it as such

ATFGK committed:
Merge branch 'general-devel' into general-devel

punchready commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2912,17 +2914,33 @@ private static bool HandleProjectileNew(GetDataHandlerArgs args)
+			if (projUUID >= 1000) projUUID = -1;
+			ai[3] = (bitsByte2[0] ? args.Data.ReadSingle() : 0f);
+			
+			//old
+			//NewProjectileData bits = new NewProjectileData((BitsByte)args.Data.ReadByte());
+			//float[] ai = new float[Projectile.maxAI];
+			//for (int i = 0; i < Projectile.maxAI; ++i)
+				//ai[i] = !bits.AI[i] ? 0.0f : args.Data.ReadSingle();
+			//ushort bannerId = bits.HasBannerIdToRespondTo ? args.Data.ReadUInt16() : (ushort)0;
+			//short dmg = bits.HasDamage ? args.Data.ReadInt16() : (short)0;
+			//float knockback = bits.HasKnockback ? args.Data.ReadSingle() : 0.0f;
+			//short origDmg = bits.HasOriginalDamage ? args.Data.ReadInt16() : (short)0;
+			//short projUUID = bits.HasUUUID ? args.Data.ReadInt16() : (short)-1;
+			//if (projUUID >= 1000)
+				//projUUID = -1;
```
no need to keep the old code, that's what git is for

ATFGK committed:
update

punchready commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2912,18 +2914,21 @@ private static bool HandleProjectileNew(GetDataHandlerArgs args)
 			short type = args.Data.ReadInt16();
-			NewProjectileData bits = new NewProjectileData((BitsByte)args.Data.ReadByte());
+			
+			BitsByte bitsByte = args.Data.ReadByte();
+		        BitsByte bitsByte2 = (BitsByte)(bitsByte[2] ? data.ReadByte() : 0);
```
just do `(byte)args.Data.ReadByte()` like the old code also did :)

ATFGK committed:
Fix errors caused by copy....

ATFGK committed:
update HandlePlayerZone

ATFGK committed:
update more

HandlePaintTile
HandlePaintWall
HandlePlaceObject
HandleNpcTeleportPortal
HandlePlayerDamageV2

ATFGK committed:
fix

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -4148,7 +4161,7 @@ private static bool HandlePlayerDamageV2(GetDataHandlerArgs args)
 			var crit = bits[0];
 			var pvp = bits[1];
-
+                        var cooldownCounter = (sbyte)args.Data.ReadInt8();
```
<img width="913" alt="Screenshot 2022-11-06 at 23 15 21" src="https://user-images.githubusercontent.com/532647/200247478-e9d54f1d-73d0-4b5f-8383-474ad53bf15d.png">

Can you fix this indentation?

hakusaro replied:
Same thing as the other ones -- please add this to the hook if you're going to read more data off.

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -1033,12 +1033,16 @@ public class PlayerZoneEventArgs : GetDataHandledEventArgs
 			/// </summary>
 			public BitsByte Zone4 { get; set; }
+			/// <summary>
+			/// ?
```
Can you document this field? `?` is not sufficient documentation for this field, unfortunately.

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2687,6 +2692,8 @@ private static bool HandleSpawn(GetDataHandlerArgs args)
 			short spawny = args.Data.ReadInt16();
 			int respawnTimer = args.Data.ReadInt32();
+			int numberOfDeathsPVE = args.Data.ReadInt16(); 
+                        int numberOfDeathsPVP = args.Data.ReadInt16(); 
```
Can you fix this indentation?

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -2912,18 +2919,21 @@ private static bool HandleProjectileNew(GetDataHandlerArgs args)
-			NewProjectileData bits = new NewProjectileData((BitsByte)args.Data.ReadByte());
+			
+			BitsByte bitsByte = (BitsByte)args.Data.ReadByte();
+		        BitsByte bitsByte2 = (BitsByte)(bitsByte[2] ? args.Data.ReadByte() : 0);
```
Can you fix this indentation?

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3150,8 +3160,8 @@ private static bool HandlePlayerZone(GetDataHandlerArgs args)
 			BitsByte zone4 = args.Data.ReadInt8();
-
-			if (OnPlayerZone(args.Player, args.Data, plr, zone1, zone2, zone3, zone4))
+                        BitsByte zone5 = args.Data.ReadInt8();
```
Can you fix this indentation?

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3649,6 +3660,7 @@ private static bool HandlePaintWall(GetDataHandlerArgs args)
 			var x = args.Data.ReadInt16();
 			var y = args.Data.ReadInt16();
 			var t = args.Data.ReadInt8();
+			var cw = args.Data.ReadInt8();//PaintCoatWall
```
You're introducing new data into this packet, but you're not actually doing anything with it. This is never used. Can you add this to the hook if you're going to read more data? e.g., `OnPaintWall(args.Player, args.Data, x, y, t, cw)`?

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3602,6 +3612,7 @@ private static bool HandlePaintTile(GetDataHandlerArgs args)
 			var x = args.Data.ReadInt16();
 			var y = args.Data.ReadInt16();
 			var t = args.Data.ReadInt8();
+			var ct = args.Data.ReadInt8();//PaintCoatTile
```
Same thing as the other ones -- please add this to the hook if you're going to read more data off.

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3914,6 +3926,7 @@ private static bool HandlePlaceObject(GetDataHandlerArgs args)
 			short type = args.Data.ReadInt16();
 			short style = args.Data.ReadInt16();
 			byte alternate = args.Data.ReadInt8();
+			sbyte random = (sbyte)args.Data.ReadInt8();
```
Same thing as the other ones -- please add this to the hook if you're going to read more data off.

Also is this really random? What?

ATFGK replied:
I don't know what this is for. It's just a new one......

ATFGK replied:
Probably related to "Rubblemaker"

hakusaro commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -4054,7 +4067,7 @@ private static bool HandlePlayerPortalTeleport(GetDataHandlerArgs args)
 		private static bool HandleNpcTeleportPortal(GetDataHandlerArgs args)
 		{
-			var npcIndex = args.Data.ReadByte();
+			var npcIndex = args.Data.ReadUInt16();
```
Same thing as the other ones -- please add this to the hook if you're going to read more data off.

hakusaro reviewed (changes requested):
Can you...

1. Implement new data in existing hooks if you read the data of fhere?
2. Document all new fields?
3. Fix indentation (we indent using tabs instead of spaces, yes this sucks)?
4. Update the changelog so that other plugin developers know what hooks you've changed?

ATFGK committed:
Update GetDataHandlers.cs

ATFGK committed:
Update and rename docs/changelog.md to 文档/changelog.md

ATFGK committed:
Update changelog.md

ATFGK committed:
Update GetDataHandlers.cs

ATFGK committed:
Fix Spawn related problems

ATFGK committed:
Fix Spawn related problems

ATFGK committed:
Update changelog.md

ATFGK committed:
Update TSPlayer.cs

hakusaro commented:
@ATFGK this is mostly good but it's probably not going to make the next release (5.1) due to Terraria update. I'll do a re-review soon. Thank you for your contributions -- they **will** get out eventually.

ATFGK committed:
Delete 文档 directory
